### PR TITLE
feature: Allow auto-normalisation of branch name

### PR DESCRIPTION
**Pull request by RussKie**
Git spec imposes a number of rules on how branches are named.
The rules can be found at https://www.git-scm.com/docs/git-check-ref-format/1.8.2

The feature allows to automatically check for compliance when the user
creates a new branch (by typing in or inserting from the clipboard).

The FormCreateBranch UI is reworked to provide a toggle for the option.

------------

**Existing dialog:**
image

**New dialog:**
Attempting to create a branch with name `[BUG-1234] some long description located [//test/*/dir?/.lock]{3,} and <foo:bar\\can>...` yeilds the following result:
image

### CODE REVIEW

**Comment by Radrik5** [comment]
1. Why do you need a checkbox? Branch creation fails if the branch name is not normalized, so maybe we should always normalize it? Or warn the user in a message box and suggest to normalize the branch name?
2. Please add an issue: https://github.com/gitextensions/gitextensions/wiki/Project-Workflow#contributing-to-the-project
3. Label of the last checkbox is truncated.

**Comment by RussKie** [comment]
1. Fair comment, the only reason I added a checkbox to allow users to contol it. Heck, out of so many thousands of users there may be diehards not wanting the feature :)
I'll remove the checkbox and make the normalisation automatic. 
There can also be added an option to the Settings for users to configure their preferred replacement token (currently set to `_`) but I haven't had any time to find my way around that part of the codebase. That option may compliment this change but is not strictly necessary.

2. Will do.

3. I missed that, thanks. I'll fix the layout.

**Comment by jbialobr** [comment]
> Fair comment, the only reason I added a checkbox to allow users to contol it. Heck, out of so many thousands of users there may be diehards not wanting the feature :)

It is always nice to have an ability to disable such a feature just in case it doesn't meet the user's expectations.

>I've removed the checkbox and made the normalisation automatic. 
There can also be added an option to the Settings for users to configure their preferred replacement token (currently set to _) but I haven't had any time to find my way around that part of the codebase. 

Have a look at [Detailed settings page](https://github.com/gitextensions/gitextensions/blob/7feb718355768072e90e1119a396b11a003aedab/GitUI/CommandsDialogs/SettingsDialog/Pages/DetailedSettingsPage.cs#L42)

**Comment by RussKie** [comment]
Added the configuration, which can be turned on/off.
Default: *on*, replacement symbol: `_`

image

**Comment by jbialobr** [comment]
Would it be better to remove diacritics from a branch name instead replacing them with _?
Eg. błąd -> blad

**Comment by RussKie** [comment]
This has actually never crossed my mind since English doesn't have those.
I've submitted this feature purely to save me (and other too) from manually replacing ` ` with `_`

My work can certainly be expanded and built upon to cater for diacritics though I am afraid at present I don't have capacity to do it myself.

**Comment by RussKie** [comment]
A quick search yielded this, which may not be 100% perfect but probably good enough 
http://stackoverflow.com/questions/249087/how-do-i-remove-diacritics-accents-from-a-string-in-net

**Comment by KindDragon** [comment]
I don't labels right alignment, all our dialog use left alignment. It's look inconsistent with others WinForms

**Comment by RussKie** [comment]
Np

Which do you prefer the most?
1. image or
2. image

Personally I think 1 looks tidier...

**Comment by jbialobr** [comment]
> I am afraid at present I don't have capacity to do it myself.

After reading the stackoverflow question, I see that the problem is more complex than I originally thought. However the rules don't forbid accented characters, so the condition `(c > 39 && c < 177)` is too restrictive.

**Comment by RussKie** [comment]
Respectfully disagree. 
As per the [spec](https://www.git-scm.com/docs/git-check-ref-format/1.8.2):

> They cannot have ASCII control characters (i.e. **bytes whose values are lower than \040, or \177 DEL**), space, tilde ~, caret ^, or colon : anywhere.

I am unable to create a branch using non-ascii characters in either cmd or git bash...

**Comment by jbialobr** [comment]
As I understand the spec, the condition should be:
1. No chars lower than 40
2. No DEL -127 (I believe  that \177 is a mistake)
3. No space, tilde ~, caret ^, or colon :
And that is it, no more chars are disallowed in the rule 4.

**Comment by jbialobr** [comment]
>I am unable to create a branch using non-ascii characters in either cmd or git bash...

I was able to create such a branch from bash: 
`git branch "błąd" "e57322e16d265a1a43c6a7c8ddfc5a02c73c309b"`
`git push jbialobr błąd`
https://github.com/jbialobr/gitextensions/tree/b%C5%82%C4%85d

**Comment by RussKie** [comment]
I take it back, on another computer I could too create a non ascii branch...

**Comment by RussKie** [comment]
> However the rules don't forbid accented characters, so the condition (c > 39 && c < 177) is too restrictive.

Raised an issue for this https://github.com/gitextensions/gitextensions/issues/3301

**Comment by RussKie** [comment]
> I don't labels right alignment, all our dialog use left alignment. It's look inconsistent with others WinForms

Raised an issue for this https://github.com/gitextensions/gitextensions/issues/3302

File: GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.resx [hunk 1]
```diff
@@ -128,4 +128,12 @@ If no, redirects select strings from stdout/stderr into an edit box. This is the
 
 Turn off if you experience problems with the console emulator.</value>
   </data>
+  <data name="chkAutoNormaliseBranchName.ToolTip" xml:space="preserve">
```
Comment (EbenZhang) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.resx]: the value is wrong, looks copied from conemu.

Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.resx]: indeed, not sure how that happened...
thanks

Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/AdvancedSettingsPage.resx]: Raised #3307

**Pull request merged**